### ISSUE TITLE
Port twisted.words.xish to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8337]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#8337|
|Type|enhancement|
|Created|2016-05-18 12:56:25Z|
|Branch|https://github.com/twisted/twisted/tree/py3-xish-8337|

<details><summary>Searchable metadata</summary>

```
trac-id__8337 8337
type__enhancement enhancement
reporter__ralphm ralphm
priority__normal normal
milestone__Python_3_x Python-3.x
branch__py3_xish_8337 py3-xish-8337
branch_author__ralphm ralphm
status__closed closed
resolution__fixed fixed
component__words words
keywords__None None
time__1463576185547428 1463576185547428
changetime__1466078289518541 1466078289518541
version__None None
owner__ralph ralph
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set owner to @ralphm|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm removed owner|
|-|-|

Created a [pull request](https://github.com/twisted/twisted/pull/72).

The bulk of the change is to `xpathparser.g` from which `xpathparser.py` is generated by Yapps. I upgraded Yapps to the [fork by Matthias Urlichs](https://github.com/smurfix/yapps). This fork is [recommended](http://theory.stanford.edu/~amitp/yapps/) by the original author, and also the source of the most recent versions over at PyPI.

The current version of this fork has been mostly ported to Python 3. There are some minor uses of `print` that don't yield a syntax error (because they are followed by the `>>` operator). There is a [pull request](https://github.com/smurfix/yapps/pull/2) to fix the remaining issues. I chose to avoid these code paths entirely by not using the generated `parse` which calls `print_line_with_pointer`. Instead of printing errors, instantiating `XPathQuery` with a malformed query string will now raise an exception that can be inspected.

As before, I imported the runtime for Yapps into `xpathparser.g`. This time because of the above. If the pull request mentioned above is accepted, and a new package appears on PyPI, and the Debian/Ubuntu port of this package doesn't introduce incompatibilities, we might opt to depend on Yapps for its runtime instead.

Another noteworthy change is that this port also fixes [#3579](https://github.com/twisted/twisted/issues/3579) (some more details there). Because of this, in Python 2 `str(element)` will return the UTF-8 encoded version of `unicode(element)`.

Further it mostly fixes [#2490](https://github.com/twisted/twisted/issues/2490). The remaining bit would be allowing Yapps to parse XPath queries that have non-ASCII code points for element and attribute name references. This requires Yapps itself to support specifying such productions, which it doesn't at the moment. For the typical use case of processing XMPP, this is likely not needed. At least all protocol extensions by the XSF use ASCII names and namespaces.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_5"></a>|@rodrigc commented|
|-|-|

I agree with the approach of using the fork from Matthias Ulrichs.  There is no easier way to move forward with this code, and pull in Python 3 fixes without going with the fork.
As you point out, even the original author of yapps2 recommends this fork.

That was the approach I was looking at when I was looking into fixing [ticket:8350].
I'm going to mark [ticket:8350] a duplicate of this one, since you have solved all the problems mentioned in that ticket.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @ralphm|
|-|-|

Thanks for the work.

I have ignored twisted/words/xish/xpathparser.py and .g since it is autogenerated ... I am happy to see that the new code result in less pyflake errors.

1. For twisted/python/compat.py maybe we should add inter to `__all__`

2. For test_characterDataNativeString can we have an assertion for the data type? I think that assertEqual is sloppy and will consider str equal to unicode if they have the same value.

3. Can we also have test_addElementContent where non-ascii values are used ... and similar tests with non-ascii data for addChild and addContent

4. What is the difference between test_characterData and test_characterDataUnicode ?

5. In twisted/words/test/test_xpath.py the docstring for test_locationWithValueUnicode does not add to much new information. I saw that you have extended the docstring for `test_queryForString` which is great. Please try to write a better docstring for locationWithValueUnicode

6. In test_queryForStringList is `"/foo"` supposed to be unicode or native string ? Is is not clear from the test's docstring

7. Do we have a test for IElement.toXML using a non-ascii defaultUri ?

8. In twisted/words/xish/domish.py for `Element.__str__`, why not always use `__unicode__` ? I think that the code needs a docstring and the docstring of the tests should be updated to explain why we can not have the same behaviour on both py2 and py3

9. In twisted/words/xish/utility.py why do we need `for key, (methodwrapper, onetime) in list(self.callbacks.items()):` this looks like just an iteration and it should be fine to have a list on py2 and iteration on py3 as they will have the same behaviour

10. `priorities = list(observers.keys())` is ok but instead of priorities.sort()

11. In twisted/words/xish/xpath.py why do we need to inline / local import for twisted.words.xish.xpathparser

Minor comments

1. Instead of `self.assertTrue(exc.msg.startswith("Trying to find one of"))` maybe we should update trial to have an self.assertStartsWith('somethig', exc.msg)` for TDD on errors assertTrue will just provide a blank `false is not true` message, while for assertStartsWith we can gave a better error reporting 

-------

Please see my comments and resubmit for review.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm removed owner|
|-|-|

Replying to [adiroiban](#note_6):
> Thanks for the work.
> 
> I have ignored twisted/words/xish/xpathparser.py and .g since it is autogenerated ... I am happy to see that the new code result in less pyflake errors.

Thanks. I'm sure it could be improved further, but that would require modifications to yapps2 itself. I don't feel there's much gain in that.

> 1. For twisted/python/compat.py maybe we should add inter to `__all__`

Fixed
 
> 2. For test_characterDataNativeString can we have an assertion for the data type? I think that assertEqual is sloppy and will consider str equal to unicode if they have the same value.

Added.
 
> 3. Can we also have test_addElementContent where non-ascii values are used ... and similar tests with non-ascii data for addChild and addContent

I redid `addContent` quite a bit through a new helper `_coerceUnicode` that mimics the old behaviour of wrapping a byte string in a `unicode` call. More details in the docstring. I've already completed my port of `twisted.words.protocols.jabber` (tickets to follow) and have used it there a couple of times, too. The new tests should cover all the cases. Note that this will now emit a warning if byte strings are used in Python 2, to make it explicit that you should really use unicode.

> 4. What is the difference between test_characterData and test_characterDataUnicode ?

The former is there to make sure the result is not a byte string when only ASCII characters are used.

> 5. In twisted/words/test/test_xpath.py the docstring for test_locationWithValueUnicode does not add to much new information. I saw that you have extended the docstring for `test_queryForString` which is great. Please try to write a better docstring for locationWithValueUnicode

Fixed. That was a copy/paste left-over.

> 6. In test_queryForStringList is `"/foo"` supposed to be unicode or native string ? Is is not clear from the test's docstring

The way the parser is generated using yapps, and how its port to Python 3 is done, the strings are used throughout for the `XPathQuery` query strings are unicode strings _or_ ASCII-only byte strings.

> 7. Do we have a test for IElement.toXML using a non-ascii defaultUri ?

Assuming you mean the `defaultUri` argument to `toXml`, it turns out that is not tested at all in `test_domish`, but indirectly through `twisted.words.protocols.jabber`. As it is only used for comparison to decide if `xmlns` should be emitted for XML Stanzas. I'm not sure if adding tests for this is in-scope for this ticket.

> 8. In twisted/words/xish/domish.py for `Element.__str__`, why not always use `__unicode__` ? I think that the code needs a docstring and the docstring of the tests should be updated to explain why we can not have the same behaviour on both py2 and py3

Ticket [#3579](https://github.com/twisted/twisted/issues/3579) explains the context of why `str`/`unicode` is used to extract the character data. This new implementation actually fixes the weird old situation, and works the same in both Python 2 and 3. What kind of docstring or explanation did you expect?

> 9. In twisted/words/xish/utility.py why do we need `for key, (methodwrapper, onetime) in list(self.callbacks.items()):` this looks like just an iteration and it should be fine to have a list on py2 and iteration on py3 as they will have the same behaviour

`self.callbacks` is modified in this loop, so it can't be an iterator.

> 10. `priorities = list(observers.keys())` is ok but instead of priorities.sort()

Instead?

> 11. In twisted/words/xish/xpath.py why do we need to inline / local import for twisted.words.xish.xpathparser

To prevent circular import issues.

> Minor comments
> 
> 1. Instead of `self.assertTrue(exc.msg.startswith("Trying to find one of"))` maybe we should update trial to have an self.assertStartsWith('somethig', exc.msg)` for TDD on errors assertTrue will just provide a blank `false is not true` message, while for assertStartsWith we can gave a better error reporting 

Agreed. For now I added a error message to this test.

Thanks!

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

Hi,

Here is a quick review. I have never used this code in production so I think that the final review should be done by someone with more experience in using this code... also in the context of [#3579](https://github.com/twisted/twisted/issues/3579) 


Here are the answers from previous review:

6. Thanks for the comment. Please try to incorporate that information into the tests.

7. My understanding is that ported py3 code need to have 100% coverage, so yes, we would need a new test here... but I might got this wrong. Please check with Glyph or Amber

11. For circular import issues, I prefer to add a comment to make it clear that the inline import is not randomly placed there... but in fact, this is a minor comment.

----------

Here are the new comments

1. I don't think that the warning from _coercedUnicode is ok... we might use this method in new placed in another release and then the version from the warning will no longer be valid.

I would say that it should not raise a deprecation warning.

2. I think that test_addContentBytes should be split into 2 tests, with a dedicated docstring for the py3 case, as I feel that the current docstring don't match the py3 behaviour

Same for test_addContentBytesNonASCII

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm commented|
|-|-|

Replying to [adiroiban](#note_8):
> Hi,
> 
> Here is a quick review. I have never used this code in production so I think that the final review should be done by someone with more experience in using this code... also in the context of [[#3579](https://github.com/twisted/twisted/issues/3579)](https://github.com/twisted/twisted/issues/3579) 
> 
> 
> Here are the answers from previous review:
> 
> 6. Thanks for the comment. Please try to incorporate that information into the tests.

I am not really sure how to add tests beyond the existing ones. The problem is that I want to test native strings here. Tests for unicode element names would currently fail because the scanner generated by yapps doesn't support that in its current form. I consider that a bug (#3579), so I don't really want to test against it.
 
> 7. My understanding is that ported py3 code need to have 100% coverage, so yes, we would need a new test here... but I might got this wrong. Please check with Glyph or Amber

Glyph mentioned in an e-mail to the mailing-list a while ago, that this only holds for lines touched by the port.

> 11. For circular import issues, I prefer to add a comment to make it clear that the inline import is not randomly placed there... but in fact, this is a minor comment.

Added.

> Here are the new comments
> 
> 1. I don't think that the warning from _coercedUnicode is ok... we might use this method in new placed in another release and then the version from the warning will no longer be valid.
> 
> I would say that it should not raise a deprecation warning.
> 
> 2. I think that test_addContentBytes should be split into 2 tests, with a dedicated docstring for the py3 case, as I feel that the current docstring don't match the py3 behaviour
> 
> Same for test_addContentBytesNonASCII

So I redid `_coercedUnicode` to remove the deprecation warning and then reworked the `addContent` tests to be more in line with what you suggested.

Additionally, I fixed some `twistedchecker` warnings. Unfortunately, the last few _new_ messages are from the generated `xpathparser` code, because of https://github.com/twisted/twistedchecker/issues/110.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_10"></a>|@ralphm commented|
|-|-|

So it turns out that twistedchecker honors `pylint: disable=`, so I silenced the erroneous warnings for `xpathparser.py`. Now the final builder is green. Yay.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

I am trying to review this, but I have no idea of how this code is supposed to be used... Also docs/words does not help to much :(

Why is xish in twisted.words ? looks like a generic XML thing, not strictly related to any text messaging ... so the whole xml thing looks to me like a twisted.xml

Please note that I have never realy used xml handling :)

Is there any tutorial for domish ?
Why is it better than stdlib xml with defusedxml ?

Is there any common thing between twisted.words.xish.xmlstream and twisted.words.xish.domish


In the near future (next 2 months) I will have to implement webdav support using Twisted... so maybe I will start using this, instead of stdlib

------------

so here is my basic review

1. test_bytesASCII looks like two tests in one :) ... and the other tests are 1.5 tests in one :)

I suggest to split the tests into 2 test cases, and skip each test case based on the python version.

2. The _coercedUnicode does not make to much sense to me... it is placed in compat, which is supposed to provide helpers for unified handling ... but I can see that _coercedUnicode behave different between python 2 and python 3... I don't understand why we need this :)

3. If we have test_characterDataUnicode I don't see why we also need test_characterData ... or maybe combine them in `u'\N{SNOWMAN} is here!'`

4. In the docstring of test_addContentBytesNonASCII can you please add the details why this is available only on python 2?


---------


If no other dev will review this, I think that you should get in touch with Glyph and see if is ok to merge it as it is.

If things will break, I hope that more people using it will start to volunteer for maintaining it.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_12"></a>|@ralphm commented|
|-|-|

Replying to [adiroiban](#note_11):
> I am trying to review this, but I have no idea of how this code is supposed to be used... Also docs/words does not help to much :(

The xish module is for working with so-called XML Streams, the base of the Jabber/XMPP protocol. The idea is that over a single (TCP) connection, pieces of XML are exchanged in a bidirectional fashion. Over the course of the lifetime of the connection, logically two XML documents are exchanged, one in each direction, with the unit of communication being the child elements of the root element.


> Why is xish in twisted.words ? looks like a generic XML thing, not strictly related to any text messaging ... so the whole xml thing looks to me like a twisted.xml

So while it could be used for other things, xish is primarily a support module for `twisted.words.protocols.jabber`. It used to be in `twisted.xish` but was [moved](https://github.com/twisted/twisted/commit/b4d1933a0a546dc305b6494c6029e83ea3cd2213) to words back in 2005.

> Please note that I have never realy used xml handling :)
> 
> Is there any tutorial for domish ?

Not at this time, unfortunately.

> Why is it better than stdlib xml with defusedxml ?

XMPP has very specific XML handling requirements. There is more elaborate description in [#3201](https://github.com/twisted/twisted/issues/3201) if you have some time to spend. There are some things that need improving, and Glyph and I have had a discussion on the various way we might or might not use `defusedxml`. Domish currently depends on either Expat or `twisted.web.sux` for parsing. The latter has not been ported to Python 3, so far.
 
> Is there any common thing between twisted.words.xish.xmlstream and twisted.words.xish.domish

Yes, `xmlstream.XmlStream` uses `domish.elementStream` to parse incoming XML in a streaming fashion and uses the DOM-ish (!) objects it yields to pass on to handlers using `utility.EventDispatcher`. Events dispatched can either have a string label, or an Element. For the latter, you can add observers to the dispatcher based in XPath-ish expressions. So it is all very XML-ish, hence the name xish.

> In the near future (next 2 months) I will have to implement webdav support using Twisted... so maybe I will start using this, instead of stdlib

There used to be WebDAV support in Twisted, based on the`web2` module. While `web2` was abandoned, it and the WebDAV implementation live on in Apple's [Calendar and Contact Server](http://calendarserver.org/). You can ask wsanchez and glyph all about it.
 
> ------------
> 
> so here is my basic review
> 
> 1. test_bytesASCII looks like two tests in one :) ... and the other tests are 1.5 tests in one :)
> 
> I suggest to split the tests into 2 test cases, and skip each test case based on the python version.

Sure, I'll work on this.
 
> 2. The _coercedUnicode does not make to much sense to me... it is placed in compat, which is supposed to provide helpers for unified handling ... but I can see that _coercedUnicode behave different between python 2 and python 3... I don't understand why we need this :)

I hoped the `_coercedUnicode` docstring explained this a bit. It is meant to replace calls of `unicode` around string-like objects, with the expectation that ASCII-only byte strings are converted to unicode strings. I didn't want to propagate this behavior to Python 3, as everything should just use the native unicode string instead there.
 
> 3. If we have test_characterDataUnicode I don't see why we also need test_characterData ... or maybe combine them in `u'\N{SNOWMAN} is here!'`

This latter test is there because there used to be an obscure bug where ASCII-only unicode strings would yield a bytes object if you called `unicode` on the element. Yeah.
 
> 4. In the docstring of test_addContentBytesNonASCII can you please add the details why this is available only on python 2?

Ok.

> ---------
>  
> If no other dev will review this, I think that you should get in touch with Glyph and see if is ok to merge it as it is.
> 
> If things will break, I hope that more people using it will start to volunteer for maintaining it.

ACK.

Thanks for reviewing. I'll get on the changes tomorrow.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban set owner to ralph|
|-|-|

Hi,

Then for 3. I think that you should update the test docstring to better explain why we have that test and what it really tries to test :)

Please get in contact with Glyph and see if you can land this.

Removing it from the review queue for now.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_14"></a>|@ralphm commented|
|-|-|

I addressed the review comments regarding the tests and their docstrings and merged in current trunk to resolve a merge conflict.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

From irc

```
(10:36:37) glyph: adiroiban: there may be nobody who knows how it should behave though.  ralphm is probably the only one
```

so, ralphm. I am trusting you so please merge this :)

When I will get to chance to look at webdav, I will give this code a try to see if can be used for parsing huge XML files in a safe stream manner with minimal memory usage :)

thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_16"></a>|@ralphm set status to `closed`|
|-|-|

In changeset 43721efe0b48729a8c2c1335947c6ad7981f0426
```
#!CommitTicketReference repository="" revision="43721efe0b48729a8c2c1335947c6ad7981f0426"
Merge py3-xish-8337: Port twisted.words.xish to Python 3.

Author: ralphm
Reviewer: adiroiban
Fixes: #8337, #3579
```
